search/api.py

```diff
@@ -301,3 +319,85 @@ def document_needs_updating(enrollment):
     serialized_enrollment = serialize_program_enrolled_user(enrollment)
     del serialized_enrollment['_id']
     return serialized_enrollment != document['_source']
+
+
+def update_percolate_memberships(user, source_type):
+    """
+    Updates membership in a PercolateQuery
+
+    Args:
+        user (User): A User to check for membership changes
+        source_type (str): The type of the percolate query to filter on
+    """
+    # ensure we have a membership for each of the queries so we can acquire a lock on them
+    percolate_queries = list(PercolateQuery.objects.filter(source_type=source_type))
+    membership_ids = _ensure_memberships_for_queries(
+        percolate_queries,
+        user
+    )
+
+    _update_memberships([query.id for query in percolate_queries], membership_ids, user)
+
+
+def _ensure_memberships_for_queries(percolate_queries, user):
+    """
+    Ensures PercolateQueryMemberships exist for the user on the designated PercolateQueries
+
+    Args:
+        percolate_queries (list of PercolateQuery): A list of PercolateQuerys to add PercolateQueryMemberships for
+        user (User): The user to ensure memberships for
+    """
+    membership_ids = []
+    for query in percolate_queries:
+        membership, _ = PercolateQueryMembership.objects.get_or_create(query=query, user=user)
+        membership_ids.append(membership.id)
+
+    return membership_ids
+
+
+def _update_memberships(percolate_query_ids, membership_ids, user, force_save=False):
+    """
+    Atomically determine and update memberships
+
+    Args:
+        percolate_query_ids (set of int): a set of PercolateQuery.id
+        membership_ids (list of int): A list of ids for PercolateQueryMemberships to update
+            user (User): A User to check for membership changes
+        force_save (bool): True if membership saves should be force even if no change
+    """
+    if not percolate_query_ids:
```

Comment:
Actually, in this case both `percolate_query_ids` and `membership_ids` are empty. Since the method that calls this determines `membership_ids` based off `percolate_query_ids`, if the latter is empty the former always will be too.

Reply:
I'm not sure we should trust that that will always be true here. Can you add `and if not membership_ids` here to make sure both are empty?

Reply:
I think we should probably exit if either are empty, right?